drivers/modem/simcom-sim7080.c

```diff
@@ -0,0 +1,2384 @@
+	(void)memset(&dns_result_addr, 0, sizeof(dns_result_addr));
+
+	/* Currently only support IPv4. */
+	dns_result.ai_family = AF_INET;
+	dns_result_addr.sa_family = AF_INET;
+	dns_result.ai_addr = &dns_result_addr;
+	dns_result.ai_addrlen = sizeof(dns_result_addr);
+	dns_result.ai_canonname = dns_result_canonname;
+	dns_result_canonname[0] = '\0';
+
+	if (service) {
+		port = atoi(service);
+		if (port < 1 || port > USHRT_MAX)
+			return DNS_EAI_SERVICE;
+	}
+
+	if (port > 0U) {
+		if (dns_result.ai_family == AF_INET)
+			net_sin(&dns_result_addr)->sin_port = htons(port);
+	}
+
+	/* Check if node is an IP address */
+	if (net_addr_pton(dns_result.ai_family, node,
+			  &((struct sockaddr_in *)&dns_result_addr)->sin_addr) == 0) {
+		*res = &dns_result;
+		return 0;
+	}
+
+	/* user flagged node as numeric host, but we failed net_addr_pton */
+	if (hints && hints->ai_flags & AI_NUMERICHOST) {
+		return DNS_EAI_NONAME;
+	}
+
+	snprintk(sendbuf, sizeof(sendbuf), "AT+CDNSGIP=\"%s\",10,20000", node);
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler, cmd, ARRAY_SIZE(cmd), sendbuf,
+			     &mdata.sem_dns, MDM_DNS_TIMEOUT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*res = (struct zsock_addrinfo *)&dns_result;
+	return 0;
+}
+
+/*
+ * Free addrinfo structure.
+ */
+static void offload_freeaddrinfo(struct zsock_addrinfo *res)
+{
+	/* No need to free static memory. */
+	res = NULL;
+}
+
+/*
+ * DNS vtable.
+ */
+const struct socket_dns_offload offload_dns_ops = {
+	.getaddrinfo = offload_getaddrinfo,
+	.freeaddrinfo = offload_freeaddrinfo,
+};
+
+static struct net_if_api api_funcs = {
+	.init = modem_net_iface_init,
+};
+
+static bool offload_is_supported(int family, int type, int proto)
+{
+	return true;
+}
+
+static int offload_socket(int family, int type, int proto)
+{
+	int ret;
+
+	ret = modem_socket_get(&mdata.socket_config, family, type, proto);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	errno = 0;
+	return ret;
+}
+
+/*
+ * Process all messages received from the modem.
+ */
+static void modem_rx(void)
+{
+	while (true) {
+		/* Wait for incoming data */
+		k_sem_take(&mdata.iface_data.rx_sem, K_FOREVER);
+
+		mctx.cmd_handler.process(&mctx.cmd_handler, &mctx.iface);
+	}
+}
+
+MODEM_CMD_DEFINE(on_cmd_ok)
+{
+	modem_cmd_handler_set_error(data, 0);
+	k_sem_give(&mdata.sem_response);
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_error)
+{
+	modem_cmd_handler_set_error(data, -EIO);
+	k_sem_give(&mdata.sem_response);
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_exterror)
+{
+	modem_cmd_handler_set_error(data, -EIO);
+	k_sem_give(&mdata.sem_response);
+	return 0;
+}
+
+/*
+ * Handles pdp context urc.
+ *
+ * The urc has the form +APP PDP: <index>,<state>.
+ * State can either be ACTIVE for activation or
+ * DEACTIVE if disabled.
+ */
+MODEM_CMD_DEFINE(on_urc_app_pdp)
+{
+	mdata.pdp_active = strcmp(argv[1], "ACTIVE") == 0;
+	LOG_INF("PDP context: %u", mdata.pdp_active);
+	k_sem_give(&mdata.sem_response);
```

Comment:
I think you might need to set the error code if you want to give `sem_response` here?